linera-views/src/common.rs

```diff
@@ -426,36 +426,23 @@ pub trait LocalWritableKeyValueStore<E> {
     async fn clear_journal(&self) -> Result<(), E>;
 }
 
-/// Config types need to be able to return their cache size
-pub trait CacheSize {
-    /// Get the cache size of the `Config` entry.
-    fn cache_size(&self) -> usize;
-}
-
 /// Low-level trait for the administration of stores and their namespaces.
 #[trait_variant::make(AdminKeyValueStore: Send)]
-pub trait LocalAdminKeyValueStore: Sized {
-    /// The error type returned by the store's methods.
-    type Error;
```

Comment:
Why is this not an associated type anymore?

Reply:
I think this is fine as there is an introduced `pub trait LocalAdminKeyValueStore<E>: Sized` with the only change being that it takes an error type `E`.

Reply:
But to me an error type as a generic argument doesn't make sense: A type will never implement `Store<SomeError>` and `Store<SomeOtherError>`. It's the implementation that determines the error type, so it should be an associated type.

Reply:
That is correct, however, we want the error types of `ReadableKeyValueStore` and `WritableKeyValueStore` to be coherent and that provides a way to do it.

Reply:
We don't need to move this associated type to a type parameter in order to constrain it: we can write `LocalKeyValueStore: LocalReadableKeyValueStore<Error = Self::Error>`, et cetera. It may be a little more verbose but it puts much less of a type annotation burden on the caller.

Reply:
One large simplification I was talking about with Matt that might make sense in this context is to remove the `Error` associated type entirely and just use `ViewError` concretely, with a `Backend(Box<dyn std::error::Error + Send>)` variant to avoid encoding all the backend error types into `ViewError`. Throughout the codebase we already require everywhere that `Error: From<ViewError>`, which necessarily means that `ViewError` has the capability to handle these errors built in.

(There's a further refinement we can then do where we separate store errors and view errors — but that can happen later)

Reply:
That construction does not work due to cyclic dependency. The following works:
```
LocalKeyValueStore: LocalReadableKeyValueStore<Error = <Self as LocalKeyValueStore>::Error> {
  type Error: Debug;
}
```

In order to avoid putting plenty of `<K as KeyValueStore>::Error` all over the code, I though it was better to have `ReadError`, `WriteError` and `AdminError`.

Reply:
Ok, thank you. That works nicely. I pushed the change in the PR. We cannot have a ViewError right now for the storages. Maybe in the future, but not possible right now.

Reply:
There's a nice series about _type members vs type parameters_ on Scala blog website – https://typelevel.org/blog/2015/07/13/type-members-parameters.html

Highly recommend

Reply:
In the Rust sphere there's recently https://gavinleroy.com/writings/i-heart-assoc-types.html as well :)